File: src/sorting/insertionSort.cs
using System;
using Sorting;

namespace InsertionSort
{
    public class InsertionSort<T> : Sorting<T> where T : IComparable<T>
    {
        public InsertionSort()
        {
            _arr = Array.ConvertAll<string, T>(Console.ReadLine().Split(' ',
            StringSplitOptions.RemoveEmptyEntries), ParseType);
            _count = _arr.Length;
        }

        public InsertionSort(T[] input)
        {
            _arr = input;
            _count = _arr.Length;
        }

        public override void Sort()
        {
            for (int i = 1; i < _count; i++)
            {
                int idx = i - 1;
                T temp = _arr[i];
                // Move elements in array until the element is smaller than the temp.
                while (idx >= 0 && temp.CompareTo(_arr[idx]) < 0)
                {
                    _arr[idx + 1] = _arr[idx];
                    idx--;
                }
                _arr[idx + 1] = temp;
            }
        }
    }

    public static class Program
    {
        public static void Main()
        {
            var a = new InsertionSort<int>();
            a.PrintArray();
            a.Sort();
            a.PrintArray();
        }
    }
}

/*
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
*/

File: src/sorting/insertionSort.py
def insertion_sort(arr, size):
    for i in range(1, size):
        temp = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > temp:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = temp


def print_array(arr, size):
    for i in range(0, size):
        print(arr[i], end=' ')
    print()


arr = list(map(int, input().split()))
size = len(arr)

print_array(arr, size)
insertion_sort(arr, size)
print_array(arr, size)

'''
Input:
5 4 3 2 1

Output:
5 4 3 2 1
1 2 3 4 5
'''